### DISCUSSION TITLE
Is it possible to add support for "Import Users" feature?

### DISCUSSION THREAD

**Discussion by potasiyam** [kreait/firebase-php#388] [Ideas]
I was checking this project and wanted to migrate existing users of a system (with password already hashed and salted) to firebase and looks like firebase has a solution for this-https://firebase.google.com/docs/auth/admin/import-users
But looks like it is not possible with this package.

It would be really nice if we had a method for importing old users in batches just like the examples in firebase documentation.

But my concern is if it is possible? I checked the rest api section of documentation of firebase authentication but found nothing about import user endpoint. If it is not possible can you suggest me any other way to do it in php?
Alternatively I am thinking about creating new user one by one but it seems the `signUpNewUser` endpoint doesn't take any field for hashed password.

**Comment by jeromegamez** [comment]
I added this as a feature request so that it doesn't get closed automatically and so that I can investigate this further.

If it were only about batch creating users (without) passwords, you would already be able to do this but importing users with hashed passwords will certainly require changes in the SDK.

As always, I'm sorry I can't give you an ETA on when this will be implemented :)

**Comment by potasiyam** [comment]
Thank you so much. For now I settled for only importing users with social media tokens. 
Thinking about the idea of using the Java/node sdk for a simple cli for the process but not important for now.
Thank you again for making this awesome package.

**Comment by MightySepp666** [comment]
This would be a nice feature, as i would not have to setup and integrate another Admin SDK project in Node or having to install the CLI on the server. My use case is to periodically sync admin users from a prod environment into a dev/staging environment.

I already found the documentation of the REST API endpoint (which is apparently not part of the firebase docs, as potasiyam mentioned): https://cloud.google.com/identity-platform/docs/reference/rest/v1/projects.accounts/batchCreate.

It seemed that integrating this endpoint would be quite straight-forward, so i just extended the `ApiClient` to make a call to 
`POST https://identitytoolkit.googleapis.com/v1/projects/{targetProjectId}/accounts:batchCreate`. 

Sample request:
[request.txt](https://github.com/kreait/firebase-php/files/4657320/request.txt)

Unfortunately, i only receive a `400: Bad Request` response with an error message stating `INSUFFICIENT_PERMISSION`. But i can't find a reason for a missing permission, as my service account has the `editor` role with all required permissions attached. So maybe, this is not the actual error.

I also tried to add the `'https://www.googleapis.com/auth/identitytoolkit'` OAuth2 scope (which is mentioned in the docs) to the `Factory` class, but that didn't solve the error.

**Comment by TPXP** [comment]
Got it to work here, all you need is to add the authenticated header to your request:
```php
return $this->requestApi("https://identitytoolkit.googleapis.com/v1/projects/{$projectId}/accounts:batchCreate", $body, [
    'access_token_auth' => 'true',
]);
```

I'll try to make a PR
